File: openwrt/packages/luci-app-mosdns-t/root/usr/lib/lua/luci/model/cbi/mosdns_t/overview.lua
local fs = require "nixio.fs"
local sys = require "luci.sys"
local http = require "luci.http"
local uci = require "luci.model.uci".cursor()

local function init_action(action)
	return os.execute("/etc/init.d/mosdns-t " .. action .. " >/dev/null 2>&1")
end

local function package_version(name)
	local version = sys.exec("opkg list-installed " .. name .. " 2>/dev/null | awk 'NR == 1 { print $3 }'")
	return version:gsub("%s+$", "")
end

local running = init_action("running") == 0
local enabled = init_action("enabled") == 0
local webui_port = uci:get("mosdns-t", "main", "webui_port") or "9099"
local host = http.getenv("HTTP_HOST") or http.getenv("SERVER_NAME") or "127.0.0.1"
host = host:gsub(":%d+$", "")

local m = Map("mosdns-t", translate("MosDNS-T"), translate("MosDNS-T service controls and OpenWrt integration."))
m.apply_on_parse = true

local s = m:section(NamedSection, "main", "mosdns-t", translate("Status"))
s.anonymous = true

local status = s:option(DummyValue, "_status", translate("Service status"))
status.rawhtml = true
function status.cfgvalue()
	if running then
		return "<span style=\"color:green\">" .. translate("Running") .. "</span>"
	end
	return "<span style=\"color:red\">" .. translate("Stopped") .. "</span>"
end

local autostart = s:option(DummyValue, "_autostart", translate("Autostart"))
function autostart.cfgvalue()
	return enabled and translate("Enabled") or translate("Disabled")
end

local core_version = s:option(DummyValue, "_core_version", translate("Core package"))
function core_version.cfgvalue()
	return package_version("mosdns-t")
end

local luci_version = s:option(DummyValue, "_luci_version", translate("LuCI package"))
function luci_version.cfgvalue()
	return package_version("luci-app-mosdns-t")
end

local controls = m:section(TypedSection, "_controls", translate("Service controls"))
controls.anonymous = true
controls.addremove = false
function controls.cfgsections()
	return { "_controls" }
end

local start = controls:option(Button, "_start", translate("Start"))
start.inputstyle = "apply"
function start.write()
	init_action("start")
	http.redirect(http.getenv("REQUEST_URI"))
end

local stop = controls:option(Button, "_stop", translate("Stop"))
stop.inputstyle = "reset"
function stop.write()
	init_action("stop")
	http.redirect(http.getenv("REQUEST_URI"))
end

local restart = controls:option(Button, "_restart", translate("Restart"))
restart.inputstyle = "reload"
function restart.write()
	init_action("restart")
	http.redirect(http.getenv("REQUEST_URI"))
end

local enable = controls:option(Button, "_enable", translate("Enable autostart"))
enable.inputstyle = "apply"
function enable.write()
	init_action("enable")
	http.redirect(http.getenv("REQUEST_URI"))
end

local disable = controls:option(Button, "_disable", translate("Disable autostart"))
disable.inputstyle = "reset"
function disable.write()
	init_action("disable")
	http.redirect(http.getenv("REQUEST_URI"))
end

local open_webui = controls:option(Button, "_open_webui", translate("Open WebUI"))
open_webui.inputstyle = "find"
function open_webui.write()
	http.redirect("http://" .. host .. ":" .. webui_port .. "/")
end

local settings = m:section(NamedSection, "main", "mosdns-t", translate("Basic settings"))
settings.anonymous = true

local enabled_flag = settings:option(Flag, "enabled", translate("Enable"))
enabled_flag.default = "1"
enabled_flag.rmempty = false

local dnsmasq_forward = settings:option(Flag, "dnsmasq_forward", translate("DNSMasq forward"))
dnsmasq_forward.default = "1"
dnsmasq_forward.rmempty = false

local listen_port = settings:option(Value, "listen_port", translate("DNS listen port"))
listen_port.datatype = "port"
listen_port.placeholder = "5335"

local web_port = settings:option(Value, "webui_port", translate("WebUI port"))
web_port.datatype = "port"
web_port.placeholder = "9099"

local config_file = settings:option(Value, "configfile", translate("Config file"))
config_file.default = "/etc/mosdns-t/config_custom.yaml"
config_file.rmempty = false

local workdir = settings:option(Value, "workdir", translate("Working directory"))
workdir.default = "/etc/mosdns-t"
workdir.rmempty = false

function m.on_after_commit()
	if fs.access("/etc/init.d/mosdns-t") then
		init_action("restart")
	end
end

return m

File: openwrt/packages/luci-app-mosdns-t/htdocs/luci-static/resources/view/mosdns-t/overview.js
'use strict';
'require form';
'require fs';
'require poll';
'require rpc';
'require uci';
'require ui';
'require view';

const callServiceList = rpc.declare({
	object: 'service',
	method: 'list',
	params: [ 'name' ],
	expect: { '': {} }
});

function getStatus() {
	return L.resolveDefault(callServiceList('mosdns-t'), {}).then(res => {
		try {
			return !!res['mosdns-t'].instances.main.running;
		} catch (e) {
			return false;
		}
	});
}

function runAction(action) {
	return fs.exec('/etc/init.d/mosdns-t', [ action ]).then(() => {
		ui.addNotification(null, E('p', {}, _('操作已执行。')), 'info');
		window.setTimeout(() => window.location.reload(), 800);
	});
}

function getAutostart() {
	return L.resolveDefault(fs.exec('/etc/init.d/mosdns-t', [ 'enabled' ]), { code: 1 })
		.then(res => res.code === 0);
}

function parsePackageStatus(result) {
	const status = {};
	String(result && result.stdout || '').trim().split('\n').forEach(line => {
		const separator = line.indexOf('=');
		if (separator > 0)
			status[line.substring(0, separator)] = line.substring(separator + 1);
	});
	return status;
}

function getPackageStatus() {
	return L.resolveDefault(fs.exec('/usr/libexec/mosdns-t-updater', [ 'status' ]), {})
		.then(parsePackageStatus);
}

function runUpdater(action) {
	return fs.exec('/usr/libexec/mosdns-t-updater', [ action ]).then(result => {
		if (result.code !== 0)
			throw new Error(result.stderr || _('软件包操作失败'));
		ui.addNotification(null, E('p', {}, action === 'upgrade'
			? _('MosDNS-T 已升级并重启。')
			: _('软件源已刷新。')), 'info');
		window.setTimeout(() => window.location.reload(), 800);
	});
}

function writePort(sectionId, option, action, value) {
	return fs.exec('/usr/libexec/mosdns-t-settings', [ action, value ]).then(result => {
		if (result.code !== 0)
			throw new Error(result.stderr || _('端口设置失败'));
		uci.set('mosdns-t', sectionId, option, value);
	});
}

return view.extend({
	load() {
		return Promise.all([
			uci.load('mosdns-t'),
			getStatus(),
			L.resolveDefault(fs.exec('/usr/bin/mosdns-t', [ 'version' ]), {}),
			getAutostart(),
			getPackageStatus()
		]);
	},

	render(data) {
		const running = data[1];
		const version = data[2] && data[2].stdout ? data[2].stdout.trim() : '';
		const autostart = data[3];
		const packageStatus = data[4] || {};
		const updateAvailable = packageStatus.update_available === '1';
		const updateChecked = packageStatus.update_checked === '1';
		const webuiPort = packageStatus.webui_port || '9099';
		const dnsPort = packageStatus.dns_port || '5335';
		let m, s, o;
		const upgradeButton = E('button', {
			class: 'btn cbi-button cbi-button-positive',
			click: ui.createHandlerFn(this, () => runUpdater('upgrade'))
		}, _('升级 MosDNS-T'));
		upgradeButton.disabled = !updateAvailable;

		m = new form.Map('mosdns-t', _('MosDNS-T'),
			_('接管 dnsmasq 时，dnsmasq 保留 53 端口并把请求转发到 MosDNS-T；未接管时，MosDNS-T 直接在所设端口提供 DNS 服务。'));

		s = m.section(form.TypedSection);
		s.anonymous = true;
		s.render = () => E('div', { class: 'cbi-section' }, [
			E('p', { id: 'mosdns-t-status' }, [
				E('strong', {}, _('运行状态：')),
				E('span', { style: 'color:%s'.format(running ? 'green' : 'red') },
					running ? _('运行中') : _('已停止'))
			]),
			E('p', {}, [
				E('strong', {}, _('版本：')),
				E('span', {}, version || _('未知'))
			]),
			E('div', { class: 'right' }, [
				E('button', {
					class: 'btn cbi-button cbi-button-apply',
					click: ui.createHandlerFn(this, () => runAction('start'))
				}, _('启动')),
				' ',
				E('button', {
					class: 'btn cbi-button cbi-button-action',
					click: ui.createHandlerFn(this, () => runAction('restart'))
				}, _('重启')),
				' ',
				E('button', {
					class: 'btn cbi-button cbi-button-negative',
					click: ui.createHandlerFn(this, () => runAction('stop'))
				}, _('停止')),
				' ',
				E('a', {
					class: 'btn cbi-button cbi-button-positive',
					href: 'http://' + window.location.hostname + ':' + webuiPort + '/',
					target: '_blank',
					rel: 'noreferrer'
				}, _('打开 MosDNS WebUI'))
			])
		]);

		s = m.section(form.TypedSection);
		s.anonymous = true;
		s.render = () => E('div', { class: 'cbi-section' }, [
			E('p', {}, '%s %s'.format(_('已安装：'), packageStatus.core_version || _('未知'))),
			E('p', {}, updateAvailable
				? _('软件源中有新版本可用。')
				: updateChecked
					? _('当前已是最新版本。')
					: _('点击“检查更新”刷新 MosDNS-T 软件源。')),
			E('div', { class: 'right' }, [
				E('button', {
					class: 'btn cbi-button cbi-button-action',
					click: ui.createHandlerFn(this, () => runUpdater('check'))
				}, _('检查更新')),
				' ',
				upgradeButton
			])
		]);

		s = m.section(form.NamedSection, 'main', 'mosdns-t', _('基础设置'));
		s.anonymous = true;

		o = s.option(form.Flag, '_autostart', _('开机自启动'));
		o.cfgvalue = () => autostart ? '1' : '0';
		o.write = (sectionId, value) => fs.exec('/etc/init.d/mosdns-t',
			[ value === '1' ? 'enable' : 'disable' ]);
		o.remove = () => fs.exec('/etc/init.d/mosdns-t', [ 'disable' ]);
		o.rmempty = false;
		o.description = _('控制路由器启动时是否自动启动 MosDNS-T。');

		o = s.option(form.Flag, 'dnsmasq_forward', _('接管 dnsmasq 上游'));
		o.default = o.enabled;
		o.rmempty = false;
		o.description = _('启用后，dnsmasq 继续监听 53，并把上游请求发往当前 MosDNS-T DNS 端口。');

		o = s.option(form.Value, 'listen_port', _('MosDNS DNS 端口'));
		o.cfgvalue = () => dnsPort;
		o.datatype = 'port';
		o.rmempty = false;
		o.write = (sectionId, value) => writePort(sectionId, 'listen_port', 'set-dns-port', value);
		o.description = _('接管 dnsmasq 时仅监听 127.0.0.1；未接管时监听全部地址，可由局域网客户端直接访问。保存并应用后会更新 UDP/TCP 监听地址和 dnsmasq 转发目标。');

		o = s.option(form.Value, 'webui_port', _('MosDNS WebUI 端口'));
		o.cfgvalue = () => webuiPort;
		o.datatype = 'port';
		o.rmempty = false;
		o.write = (sectionId, value) => writePort(sectionId, 'webui_port', 'set-webui-port', value);
		o.description = _('与 MosDNS WebUI 使用同一个端口设置；从任一界面修改后，另一边都会显示相同端口。');

		poll.add(() => getStatus().then(isRunning => {
			const el = document.getElementById('mosdns-t-status');
			if (el)
				el.lastElementChild.textContent = isRunning ? _('运行中') : _('已停止');
		}));

		return m.render();
	}
});
